src/ray/common/ray_syncer/ray_syncer.cc

```diff
@@ -0,0 +1,436 @@
+
+#include <functional>
+
+#include "ray/common/ray_config.h"
+namespace ray {
+namespace syncer {
+
+NodeState::NodeState() { snapshots_versions_taken_.fill(-1); }
+
+bool NodeState::SetComponent(RayComponentId cid,
+                             const ReporterInterface *reporter,
+                             ReceiverInterface *receiver) {
+  if (cid < static_cast<RayComponentId>(kComponentArraySize) &&
+      reporters_[cid] == nullptr && receivers_[cid] == nullptr) {
+    reporters_[cid] = reporter;
+    receivers_[cid] = receiver;
+    return true;
+  } else {
+    RAY_LOG(FATAL) << "Fail to set components, component_id:" << cid
+                   << ", reporter:" << reporter << ", receiver:" << receiver;
+    return false;
+  }
+}
+
+std::optional<RaySyncMessage> NodeState::GetSnapshot(RayComponentId cid) {
+  if (reporters_[cid] == nullptr) {
+    return std::nullopt;
+  }
+  auto message = reporters_[cid]->Snapshot(snapshots_versions_taken_[cid], cid);
+  if (message != std::nullopt) {
+    snapshots_versions_taken_[cid] = message->version();
+    RAY_LOG(DEBUG) << "Snapshot taken: cid:" << cid << ", version:" << message->version()
+                   << ", node:" << NodeID::FromBinary(message->node_id());
+  }
+  return message;
+}
+
+bool NodeState::ConsumeMessage(std::shared_ptr<const RaySyncMessage> message) {
+  auto &current = cluster_view_[message->node_id()][message->component_id()];
+
+  RAY_LOG(DEBUG) << "ConsumeMessage: " << (current ? current->version() : -1)
+                 << " message_version: " << message->version()
+                 << ", message_from: " << NodeID::FromBinary(message->node_id());
+  // Check whether newer version of this message has been received.
+  if (current && current->version() >= message->version()) {
+    return false;
+  }
+
+  current = message;
+  auto receiver = receivers_[message->component_id()];
+  if (receiver != nullptr) {
+    receiver->Update(message);
+  }
+  return true;
+}
+
+NodeSyncConnection::NodeSyncConnection(RaySyncer &instance,
+                                       instrumented_io_context &io_context,
+                                       std::string node_id)
+    : instance_(instance), io_context_(io_context), node_id_(std::move(node_id)) {}
+
+void NodeSyncConnection::ReceiveUpdate(RaySyncMessages messages) {
+  for (auto &message : *messages.mutable_sync_messages()) {
+    auto &node_versions = GetNodeComponentVersions(message.node_id());
+    RAY_LOG(DEBUG) << "Receive update: "
+                   << " component_id=" << message.component_id()
+                   << ", message_version=" << message.version()
+                   << ", local_message_version=" << node_versions[message.component_id()];
+    if (node_versions[message.component_id()] < message.version()) {
+      node_versions[message.component_id()] = message.version();
+    }
+    instance_.BroadcastMessage(std::make_shared<RaySyncMessage>(std::move(message)));
```

Comment:
Do we need to broadcast in the else case? Doesn't that mean we received an old message that should be discarded?

Reply:
I agree with you.